pkg/codegen/util_registry.go

```diff
@@ -0,0 +1,99 @@
+package codegen
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+
+	"cuelang.org/go/cue"
+	cueformat "cuelang.org/go/cue/format"
+	"github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+)
+
+const (
+	GITHUB_OWNER = "grafana"
+	GITHUB_REPO  = "kind-registry"
+)
+
+// GetPublishedKind retrieve the latest published kind from the schema registry
+func GetPublishedKind(name string, category string, latestRegistryDir string) (string, error) {
+	ctx := context.Background()
+	tc := oauth2.NewClient(ctx, nil)
+	client := github.NewClient(tc)
+
+	if latestRegistryDir == "" {
+		return "", nil
+	}
+
+	file, _, resp, err := client.Repositories.GetContents(ctx, GITHUB_OWNER, GITHUB_REPO, fmt.Sprintf("grafana/%s/%s/%s.cue", latestRegistryDir, category, name), nil)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			return "", nil
+		}
+		return "", fmt.Errorf("error retrieving published kind from GH, %d: %w", resp.StatusCode, err)
+	}
+	content, err := file.GetContent()
+	if err != nil {
+		return "", fmt.Errorf("error decoding published kind content: %w", err)
+	}
+
+	return content, nil
+}
+
+func FindLatestDir(ctx context.Context, client *github.Client) (string, error) {
+	re := regexp.MustCompile(`([0-9]+)\.([0-9]+)\.([0-9]+)`)
+	latestVersion := []uint64{0, 0, 0}
+	latestDir := ""
+
+	_, dir, _, err := client.Repositories.GetContents(ctx, GITHUB_OWNER, GITHUB_REPO, "grafana", nil)
+	if err != nil {
+		return "", err
+	}
+
+	for _, content := range dir {
+		if content.GetType() != "dir" {
+			continue
+		}
+
+		parts := re.FindStringSubmatch(content.GetName())
+		if parts == nil || len(parts) < 4 {
+			continue
+		}
+
+		version := make([]uint64, len(parts)-1)
+		for i := 1; i < len(parts); i++ {
+			version[i-1], _ = strconv.ParseUint(parts[i], 10, 32)
+		}
+
+		if isLess(latestVersion, version) {
+			latestVersion = version
+			latestDir = content.GetName()
+		}
+	}
+
+	return latestDir, nil
+}
+
+func isLess(v1 []uint64, v2 []uint64) bool {
```

Comment:
This looks fine for now, but if we need to support other version formats (i.e. beta releases) at some point, we'll need to extend this support. So, I'm wondering if we might want to use something like https://pkg.go.dev/golang.org/x/mod/semver and get that support for "free".

Reply:
I can make this update but I don't know if we want to publish kinds for beta releases. I image having a `next` (or `latest`) folder in the kind registry if you always want to work with the most updated kinds, else users will probably target a specific stable version?